### ISSUE TITLE
Flaky test: `GroupBox_FlatStyle_VisualStyles_on_SetWithCustomOldValue_GetReturnsExpected` with visual styles on

### ISSUE THREAD

**Issue by RussKie** [dotnet/winforms#4213]
The test executed from VS returns all green, run from a cli however the test fails:
```
    System.Windows.Forms.Tests.GroupBoxTests.GroupBox_FlatStyle_VisualStyles_on_SetWithCustomOldValue_GetReturnsExpected(oldValue: Popup, value: System, containerControl: True, ownerDraw: False, userMouse: False, expectedInvalidatedCallCount: 1, expectedCreatedCallCount: 1) [FAIL]
      Assert.Equal() Failure
      Expected: 1
      Actual:   0
      Stack Trace:
        C:\Development\winforms\src\System.Windows.Forms\tests\UnitTests\System\Windows\Forms\GroupBoxTests.cs(879,0): at System.Windows.Forms.Tests.GroupBoxTests.GroupBox_FlatStyle_VisualStyles_on_SetWithCustomOldValue_GetReturnsExpected(FlatStyle oldValue, FlatStyle value, Boolean containerControl, Boolean ownerDraw, Boolean userMouse, Int32 expectedInvalidatedCallCount, Int32 expectedCreatedCallCount)
    System.Windows.Forms.Tests.GroupBoxTests.GroupBox_FlatStyle_VisualStyles_on_SetWithCustomOldValue_GetReturnsExpected(oldValue: Standard, value: System, containerControl: True, ownerDraw: False, userMouse: False, expectedInvalidatedCallCount: 1, expectedCreatedCallCount: 1) [FAIL]
      Assert.Equal() Failure
      Expected: 1
      Actual:   0
      Stack Trace:
        C:\Development\winforms\src\System.Windows.Forms\tests\UnitTests\System\Windows\Forms\GroupBoxTests.cs(879,0): at System.Windows.Forms.Tests.GroupBoxTests.GroupBox_FlatStyle_VisualStyles_on_SetWithCustomOldValue_GetReturnsExpected(FlatStyle oldValue, FlatStyle value, Boolean containerControl, Boolean ownerDraw, Boolean userMouse, Int32 expectedInvalidatedCallCount, Int32 expectedCreatedCallCount)
```

Relates to https://github.com/dotnet/winforms/pull/4197

**Comment by weltkante** [comment]
```
(oldValue: Popup, value: System, containerControl: True, ownerDraw: False, userMouse: False, expectedInvalidatedCallCount: 1, expectedCreatedCallCount: 1)
(oldValue: Standard, value: System, containerControl: True, ownerDraw: False, userMouse: False, expectedInvalidatedCallCount: 1, expectedCreatedCallCount: 1)
```
I don't see these cases in the source, and the parameters logged in this issue seem wrong, you shouldn't expect both invalidation and handle recreation, thats not what the implementation does. The fact that the implementation does a `RecreateHandle` and the control gets invalidated anyways is a strong indicator that you have no control over invalidation and shouldn't be testing it.

The flakyness seems to come from an [`WM_UPDATEUISTATE`](https://docs.microsoft.com/en-us/windows/win32/menurc/wm-updateuistate) which triggers the additional invalidation instead of doing a plain `RecreateHandle`.

<details><summary>Stack trace of the unexpected invalidation</summary>

```
System.Windows.Forms.Tests.dll!System.Windows.Forms.Tests.GroupBoxTests.GroupBox_FlatStyle_VisualStyles_on_SetWithCustomOldValue_GetReturnsExpected.AnonymousMethod__0(object sender, System.Windows.Forms.InvalidateEventArgs e) Line 865	C#
System.Windows.Forms.dll!System.Windows.Forms.Control.OnInvalidated(System.Windows.Forms.InvalidateEventArgs e) Line 7962	C#
System.Windows.Forms.dll!System.Windows.Forms.Control.NotifyInvalidate(System.Drawing.Rectangle invalidatedArea) Line 6896	C#
System.Windows.Forms.dll!System.Windows.Forms.Control.Invalidate(bool invalidateChildren) Line 6223	C#
System.Windows.Forms.dll!System.Windows.Forms.Control.WmUpdateUIState(ref System.Windows.Forms.Message m) Line 12644	C#
System.Windows.Forms.dll!System.Windows.Forms.Control.WndProc(ref System.Windows.Forms.Message m) Line 12986	C#
System.Windows.Forms.dll!System.Windows.Forms.GroupBox.WndProc(ref System.Windows.Forms.Message m) Line 708	C#
System.Windows.Forms.dll!System.Windows.Forms.Control.ControlNativeWindow.OnMessage(ref System.Windows.Forms.Message m) Line 67	C#
System.Windows.Forms.dll!System.Windows.Forms.Control.ControlNativeWindow.WndProc(ref System.Windows.Forms.Message m) Line 119	C#
System.Windows.Forms.dll!System.Windows.Forms.NativeWindow.Callback(System.IntPtr hWnd, Interop.User32.WM msg, System.IntPtr wparam, System.IntPtr lparam) Line 372	C#
[Native to Managed Transition]	
[Managed to Native Transition]	
System.Windows.Forms.dll!System.Windows.Forms.NativeWindow.DefWndProc(ref System.Windows.Forms.Message m) Line 515	C#
System.Windows.Forms.dll!System.Windows.Forms.Control.DefWndProc(ref System.Windows.Forms.Message m) Line 5031	C#
System.Windows.Forms.dll!System.Windows.Forms.Control.WmUpdateUIState(ref System.Windows.Forms.Message m) Line 12577	C#
System.Windows.Forms.dll!System.Windows.Forms.Control.WndProc(ref System.Windows.Forms.Message m) Line 12986	C#
System.Windows.Forms.dll!System.Windows.Forms.ScrollableControl.WndProc(ref System.Windows.Forms.Message m) Line 1454	C#
System.Windows.Forms.dll!System.Windows.Forms.ContainerControl.WndProc(ref System.Windows.Forms.Message m) Line 1952	C#
System.Windows.Forms.dll!System.Windows.Forms.Application.ParkingWindow.WndProc(ref System.Windows.Forms.Message m) Line 155	C#
System.Windows.Forms.dll!System.Windows.Forms.Control.ControlNativeWindow.OnMessage(ref System.Windows.Forms.Message m) Line 67	C#
System.Windows.Forms.dll!System.Windows.Forms.Control.ControlNativeWindow.WndProc(ref System.Windows.Forms.Message m) Line 119	C#
System.Windows.Forms.dll!System.Windows.Forms.NativeWindow.Callback(System.IntPtr hWnd, Interop.User32.WM msg, System.IntPtr wparam, System.IntPtr lparam) Line 372	C#
[Native to Managed Transition]	
[Managed to Native Transition]	
System.Windows.Forms.dll!System.Windows.Forms.NativeWindow.DefWndProc(ref System.Windows.Forms.Message m) Line 515	C#
System.Windows.Forms.dll!System.Windows.Forms.Control.DefWndProc(ref System.Windows.Forms.Message m) Line 5031	C#
System.Windows.Forms.dll!System.Windows.Forms.Control.WndProc(ref System.Windows.Forms.Message m) Line 13021	C#
System.Windows.Forms.dll!System.Windows.Forms.ScrollableControl.WndProc(ref System.Windows.Forms.Message m) Line 1454	C#
System.Windows.Forms.dll!System.Windows.Forms.ContainerControl.WndProc(ref System.Windows.Forms.Message m) Line 1952	C#
System.Windows.Forms.dll!System.Windows.Forms.Application.ParkingWindow.WndProc(ref System.Windows.Forms.Message m) Line 155	C#
System.Windows.Forms.dll!System.Windows.Forms.Control.ControlNativeWindow.OnMessage(ref System.Windows.Forms.Message m) Line 67	C#
System.Windows.Forms.dll!System.Windows.Forms.Control.ControlNativeWindow.WndProc(ref System.Windows.Forms.Message m) Line 119	C#
System.Windows.Forms.dll!System.Windows.Forms.NativeWindow.Callback(System.IntPtr hWnd, Interop.User32.WM msg, System.IntPtr wparam, System.IntPtr lparam) Line 372	C#
[Native to Managed Transition]	
[Managed to Native Transition]	
System.Windows.Forms.dll!System.Windows.Forms.NativeWindow.DefWndProc(ref System.Windows.Forms.Message m) Line 515	C#
System.Windows.Forms.dll!System.Windows.Forms.Control.DefWndProc(ref System.Windows.Forms.Message m) Line 5031	C#
System.Windows.Forms.dll!System.Windows.Forms.Control.WndProc(ref System.Windows.Forms.Message m) Line 13021	C#
System.Windows.Forms.dll!System.Windows.Forms.GroupBox.WndProc(ref System.Windows.Forms.Message m) Line 708	C#
System.Windows.Forms.dll!System.Windows.Forms.Control.ControlNativeWindow.OnMessage(ref System.Windows.Forms.Message m) Line 67	C#
System.Windows.Forms.dll!System.Windows.Forms.Control.ControlNativeWindow.WndProc(ref System.Windows.Forms.Message m) Line 119	C#
System.Windows.Forms.dll!System.Windows.Forms.NativeWindow.Callback(System.IntPtr hWnd, Interop.User32.WM msg, System.IntPtr wparam, System.IntPtr lparam) Line 372	C#
[Native to Managed Transition]	
[Managed to Native Transition]	
System.Windows.Forms.dll!System.Windows.Forms.NativeWindow.DefWndProc(ref System.Windows.Forms.Message m) Line 515	C#
System.Windows.Forms.dll!System.Windows.Forms.Control.DefWndProc(ref System.Windows.Forms.Message m) Line 5031	C#
System.Windows.Forms.dll!System.Windows.Forms.Control.WmCreate(ref System.Windows.Forms.Message m) Line 11813	C#
System.Windows.Forms.dll!System.Windows.Forms.Control.WndProc(ref System.Windows.Forms.Message m) Line 12800	C#
System.Windows.Forms.dll!System.Windows.Forms.GroupBox.WndProc(ref System.Windows.Forms.Message m) Line 708	C#
System.Windows.Forms.dll!System.Windows.Forms.Control.ControlNativeWindow.OnMessage(ref System.Windows.Forms.Message m) Line 67	C#
System.Windows.Forms.dll!System.Windows.Forms.Control.ControlNativeWindow.WndProc(ref System.Windows.Forms.Message m) Line 119	C#
System.Windows.Forms.dll!System.Windows.Forms.NativeWindow.Callback(System.IntPtr hWnd, Interop.User32.WM msg, System.IntPtr wparam, System.IntPtr lparam) Line 372	C#
[Native to Managed Transition]	
[Managed to Native Transition]	
System.Windows.Forms.Primitives.dll!Interop.User32.CreateWindowExW(Interop.User32.WS_EX dwExStyle, string lpClassName, string lpWindowName, Interop.User32.WS dwStyle, int X, int Y, int nWidth, int nHeight, System.IntPtr hWndParent, System.IntPtr hMenu, System.IntPtr hInst, object lpParam) Line 43	C#
System.Windows.Forms.dll!System.Windows.Forms.NativeWindow.CreateHandle(System.Windows.Forms.CreateParams cp) Line 460	C#
System.Windows.Forms.dll!System.Windows.Forms.Control.CreateHandle() Line 4925	C#
System.Windows.Forms.dll!System.Windows.Forms.Control.RecreateHandleCore() Line 9622	C#
System.Windows.Forms.dll!System.Windows.Forms.Control.RecreateHandle() Line 9544	C#
System.Windows.Forms.dll!System.Windows.Forms.GroupBox.FlatStyle.set(System.Windows.Forms.FlatStyle value) Line 211	C#
System.Windows.Forms.Tests.dll!System.Windows.Forms.Tests.GroupBoxTests.GroupBox_FlatStyle_VisualStyles_on_SetWithCustomOldValue_GetReturnsExpected(System.Windows.Forms.FlatStyle oldValue, System.Windows.Forms.FlatStyle value, bool containerControl, bool ownerDraw, bool userMouse, int expectedInvalidatedCallCount, int expectedCreatedCallCount) Line 873	C#
```

</details>

The message chain on the stack is `WM_CREATE -> WM_CHANGEUISTATE -> [recurse to root HWND and back] -> WM_UPDATEUISTATE`. The [`WM_CHANGEUISTATE`](https://docs.microsoft.com/en-us/windows/win32/menurc/wm-changeuistate) message is sent with `UIS_INITIALIZE` as you would expect for a newly created control.

> If the low-order word of wParam is UIS_INITIALIZE, the system will send the WM_UPDATEUISTATE message with a UI state based on the last input event. For example, if the last input came from the mouse, the system will hide the keyboard cues. And, if the last input came from the keyboard, the system will show the keyboard cues. If the state that results from processing WM_CHANGEUISTATE is the same as the old state, DefWindowProc does not send this message.

This means whether or not you see the additional invalidation depends on the previous input state, which of course is a global state. If the last input was keyboard input you get _no invalidation_ - if the last input was mouse input you _do get invalidation_. Since you are running multithreaded tests which also sometimes simulate keyboard or mouse input the whole setup is flaky by design.

TLDR: don't test Invalidations if your tested code creates or recreates handles, you are not under control over the exact number of invalidations that are going to happen. Handle creation may or may not invalidate depending on whether the last input was keyboard or mouse.

What you can do is a `actualInvalidations >= expectedInvalidations` test, ensuring that in the case where you don't call RecreateHandle you do get at invalidated at least once. You cannot provide an upper bound however since the system may perform additional invalidations.

**Comment by RussKie** [comment]
Thank you for the assessment!

> TLDR: don't test Invalidations if your tested code creates or recreates handles, you are not under control over the exact number of invalidations that are going to happen. Handle creation may or may not invalidate depending on whether the last input was keyboard or mouse.

👍
